ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -1268,23 +1275,23 @@ type ChainAndLedger blk = ValidatedFragment (Header blk) (LedgerDB' blk)
 -------------------------------------------------------------------------------}
 
 -- | Check whether a 'ChainDiff' can be pipelined. If it can, the tentative
--- header is returned.
+-- header as well as the new 'TentativeHeaderState' (to be used in case the
+-- block body turns out to be invalid) is returned.
 --
 -- PRECONDITION: The 'ChainDiff' fits on top of the current chain and is better.
 isPipelineable ::
-     LedgerSupportsProtocol blk
+     (HasHeader (Header blk), BlockSupportsDiffusionPipelining blk)
   => BlockConfig blk
-  -> TentativeState blk
+  -> TentativeHeaderState blk
   -> ChainDiff (Header blk)
-  -> StrictMaybe (Header blk)
-isPipelineable bcfg tentativeState ChainDiff {..}
-  | -- we apply exactly one header
-    AF.Empty _ :> hdr <- getSuffix
-  , preferToLastInvalidTentative bcfg tentativeState hdr
+  -> Maybe (Header blk, TentativeHeaderState blk)
+isPipelineable bcfg st ChainDiff {..} = do
```

Comment:
very minor: I like the explicit guard syntax better. In particular, `MonadFail` always smells bad to me---especially with a `pure` at the head 🤷